## B. Normalize the data (5 marks) 

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

How does the mean of the mean squared errors compare to that from Step A?


Submit your Jupyter Notebook with your code and comments.


## Steps

Import all the libraries

In [58]:
import pandas as pd
import numpy as np
import sklearn.metrics as metrics
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense

import warnings
warnings.simplefilter('ignore', FutureWarning)

## Reading the data

In [59]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

#### 1. Randomly split the data into a training and test sets by holding 30% of the data for testing.

In [61]:
# Getting the features and targets
features = concrete_data.drop(columns='Strength', axis=1)
Y = concrete_data['Strength']

features = (features - features.mean()) / features.std()

n_cols = features.shape[1] # number of predictors

x_train, x_test, y_train, y_test = train_test_split(features, Y, test_size=0.3)


#### 2. Train the model on the training data using 50 epochs.

In [62]:
def regression_model():
    # create model
    model = Sequential()
    #One hidden layer of 10 nodes, and a ReLU activation function 
    model.add(Dense(10, activation='relu', input_shape=(n_cols,))) 
    model.add(Dense(1))
    
    # Use the adam optimizer and the mean squared error as the loss function.
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

model = regression_model()
model

#### 2. Train the model on the training data using 50 epochs.

In [63]:
model.fit(x_train, y_train, epochs=50, verbose=2)

Epoch 1/50
 - 0s - loss: 1576.3267
Epoch 2/50
 - 0s - loss: 1559.9936
Epoch 3/50
 - 0s - loss: 1544.2285
Epoch 4/50
 - 0s - loss: 1528.3905
Epoch 5/50
 - 0s - loss: 1512.2920
Epoch 6/50
 - 0s - loss: 1495.6140
Epoch 7/50
 - 0s - loss: 1478.2074
Epoch 8/50
 - 0s - loss: 1459.9721
Epoch 9/50
 - 0s - loss: 1440.3237
Epoch 10/50
 - 0s - loss: 1419.9032
Epoch 11/50
 - 0s - loss: 1397.8858
Epoch 12/50
 - 0s - loss: 1374.8616
Epoch 13/50
 - 0s - loss: 1350.5342
Epoch 14/50
 - 0s - loss: 1324.9192
Epoch 15/50
 - 0s - loss: 1298.0854
Epoch 16/50
 - 0s - loss: 1270.5690
Epoch 17/50
 - 0s - loss: 1240.9877
Epoch 18/50
 - 0s - loss: 1211.3708
Epoch 19/50
 - 0s - loss: 1180.3258
Epoch 20/50
 - 0s - loss: 1149.4572
Epoch 21/50
 - 0s - loss: 1117.0178
Epoch 22/50
 - 0s - loss: 1084.8392
Epoch 23/50
 - 0s - loss: 1052.0015
Epoch 24/50
 - 0s - loss: 1018.7609
Epoch 25/50
 - 0s - loss: 985.4326
Epoch 26/50
 - 0s - loss: 951.9779
Epoch 27/50
 - 0s - loss: 918.7926
Epoch 28/50
 - 0s - loss: 885.5512
Epoch

#### 3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength.

In [64]:
predictions = model.predict(x_test)
MAE =  metrics.mean_absolute_error(y_test, predictions)
print(MAE)

15.798754999876795


#### 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [65]:
MAEs = []
for n in range(50): 
    x_train, x_test, y_train, y_test = train_test_split(features, Y, test_size=0.3)
    model.fit(x_train, y_train, epochs=50, verbose=0)
    predictions = model.predict(x_test)
    MAEs.append(metrics.mean_absolute_error(y_test, predictions))
    
print(MAEs)    

[10.194701387349841, 9.442462728062107, 8.881074065915204, 8.325766450332592, 7.906546805174991, 7.22224119278991, 6.91054704783418, 6.653526413186082, 6.152807007576655, 6.268683466124303, 6.052534677449939, 5.773929232193042, 5.582777055413207, 5.092437293660679, 5.474677126801129, 5.433659321868304, 5.521908234778346, 5.279788445222724, 5.583608286280462, 5.464137976316186, 4.659083786134195, 5.223400725948001, 5.107630646591433, 4.884239975024967, 4.983075872871868, 4.9371457757918975, 5.607880339113254, 4.966170038636834, 4.904358647170577, 4.944357975493744, 4.9512697805633055, 5.069479215677502, 5.068636115268596, 5.120857477342426, 5.149600323797429, 5.0692069659186805, 5.3808348155716095, 5.196819080797213, 5.143390787797453, 5.121358623072553, 5.515164285307948, 4.8366695058384375, 5.2461995156297405, 5.026565174979299, 5.254655561045922, 4.919743517347911, 5.1979855546056255, 5.122334373498811, 5.137457205516235, 4.945473394609964]


#### 5. Report the mean and the standard deviation of the mean squared errors.

In [68]:
np_array = np.array(MAEs)
print(f"Mean of mean squared errors:{np.mean(np_array)} ")
print(f"Standard deviation of mean squared errors:{np.std(np_array)} ")

Mean of mean squared errors:5.718177185425865 
Standard deviation of mean squared errors:1.2205812457213703 


### How does the mean of the mean squared errors compare to that from Step A?


Step A:

- Mean of mean squared errors:5.57246862771433 
- Standard deviation of mean squared errors:0.7217945033478287 

Step B:
 - Mean of mean squared errors:5.718177185425865 
 - Standard deviation of mean squared errors:1.2205812457213703 

Not a big change